lib/compiler-llvm/src/translator/code.rs

```diff
@@ -3495,201 +3261,542 @@ impl<'ctx, 'a> LLVMFunctionCodeGenerator<'ctx, 'a> {
              * https://github.com/sunfishcode/wasm-reference-manual/blob/master/WebAssembly.md#floating-point-arithmetic-instructions
              ***************************/
             Operator::F32Add => {
-                let ((v1, i1), (v2, i2)) = self.state.pop2_extra()?;
-                let (v1, v2) = (v1.into_float_value(), v2.into_float_value());
-                let res = self.builder.build_float_add(v1, v2, "");
-                self.state.push1_extra(
-                    res,
-                    (i1.strip_pending() & i2.strip_pending()) | ExtraInfo::pending_f32_nan(),
-                );
+                let (v1, v2) = self.state.pop2()?;
+                let v1 = self.canonicalize_nans(v1);
+                let v2 = self.canonicalize_nans(v2);
+                let res = self
+                    .builder
+                    .build_call(
+                        self.intrinsics.add_f32,
+                        &[
+                            v1,
+                            v2,
+                            inkwell::values::BasicValueEnum::MetadataValue(
+                                self.intrinsics.fp_rounding_md,
```

Comment:
It'd be nice if this was named to say what the rounding mode is. Similarly for exception.